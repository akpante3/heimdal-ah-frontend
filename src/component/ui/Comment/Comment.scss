@import '../../../styles/variables.scss';

.comment-header {
  font-family: $font_family_reem-kufi;
  padding-top: 40px;
  font-size: 38px;
  text-align: center;
}
.comment__form {
  margin-bottom: 30px;
}