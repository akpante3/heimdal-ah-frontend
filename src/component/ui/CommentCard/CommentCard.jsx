import React from 'react';
import { connect } from 'react-redux';
import './CommentCard.scss';
import FontAwesome from 'react-fontawesome';
import { getArticleComment } from '../../../actions/CommentActions/CommentActions';
import DeleteButton from '../CommentForm/CommentDelete';
import ReplyComment from '../ReplyComment/ReplyComment';
import Ellipse from '../Elllipse/Ellipse';

/**
  * renderComponent
  * @method
  * @summary React component for displaying a the comment card
  * @param {object}  comment contains the comments of an article
  * @param {Object} props - React PropTypes
  * @property {String|Object} className - String className compatible object for styling
  * @return {Node} React node containing comment card view
*/

const CommentCard = (props) => {
  const { user } = props.user;
  let { comments } = props;
  comments = comments || [];
  const renderCard = () => comments.map(item => (
    <section key={item.id} className="comment-section container">
      <div className="comment-card row">
        <div className="comment-image col-md-2 col-sm-2">
          <img
            className="comment-card-img"
            src={item.profile.image}
            alt='user-comment-logo'
          />
        </div>
        <div className="comment-card-body col-md-9 col-sm-9">
          <div className="comment-body-top">
            <span className='comment-name'>{item.profile.username}</span>
            <span className='comment-date small_text'>
              <FontAwesome name='clock' className='fav_icons' />
              {item.createdAt}
            </span>
            <span className='small_text'>
              <FontAwesome name='calendar' className='fav_icons' />
              09: 00 am
            </span>
            <Ellipse articleId={item.articleId} commentId={item.id} />
            { item.userId === user.userId
              ? <DeleteButton buttonName='comment' commentId={item.id} />
              : <div />
            }
          </div>
          <div className="comment-body">
            {item.content}
          </div>
          <div className="comment-body-footer">
            <span>
              <ReplyComment commentId={item.id} />
            </span>
            <span>
              {' '}
            </span>
          </div>
        </div>
      </div>
    </section>
  ));

  /**
     * Returns the jsx of the component.
     *
     * @memberof app.components.CommentCard
     * @return {string} - HTML markup for the component
   */

  return (
    <div>
      {
      renderCard()
    }
    </div>
  );
};

/**
 * @method module:Reactator.ReduxContainerBuilderMapStateToProps
 *
 * @param {function} mapStateToProps - function mapping redux state to props
 *
 * @return {ReduxContainerBuilder} this builder
*/
const mapStateToProps = state => ({
  comments: state.comment.comments,
  user: state.auth,
});

export default connect(mapStateToProps, { getArticleComment })(CommentCard);
