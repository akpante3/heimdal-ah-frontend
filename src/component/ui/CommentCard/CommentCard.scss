@import '../../../styles/variables.scss';

.comment-section {
  font-family: $font_family_reem-kufi;
  margin-top: 50px;
  margin-bottom: 50px;

  .comment-image {
    margin-right: -10px;
  }

  .comment-card {
    border-top: 1px solid $heimdal-comment-border;
    margin: auto;
    width: 100%;
    padding-top: 20px;
  }

  .comment-body {
    margin-top: 10px;
    text-align: justify;
  }

  .comment-card-img {
    border-radius: 70px;
    height: 100px;
    margin-right: -12px;
    padding: 20px;
    width: 100px;
  }

  .comment-body-top {
    margin-top: 10px;
    padding-bottom: 22px;
  }

  .comment-name {
    margin-right: 30px;
  }

  .comment-date {
    margin-right: 30px;
  }

  .fav_icons {
    margin-right: 10px!important;
  }

  .comment-body-footer {
    margin-top: 15px;
  }

  .small_text {
    color: $heimdal-textfield-background;
    font-size: 12px;
  }

  .archive_button {
    float: right;
  }

}

