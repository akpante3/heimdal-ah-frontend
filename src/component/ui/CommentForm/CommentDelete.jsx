import React from 'react';
import FontAwesome from 'react-fontawesome';
import { connect } from 'react-redux';
import { deleteArticleComment } from '../../../actions/CommentActions/CommentActions';
import { deleteCommentReplies } from '../../../actions/ReplyActions/ReplyAction';

/**
  * renderComponent
  * @method Function Function based Component
  * @summary React component for rendering the card components
  * @param {Object} props - React PropTypes
  * @property {String|Object} className - String className compatible object for styling
  * @return {Node} React node containing delete comment and replies
  */
const DeleteButton = (props) => {
  const deleteComment = (e) => {
    e.preventDefault();
    const { commentId, buttonName, replyId } = props;
    if (buttonName === 'comment') {
      props.deleteArticleComment(1, commentId);
    } else {
      props.deleteCommentReplies(commentId, replyId);
    }
  };

  /**
     * Renders the component.
     *
     * @memberof app.components.Comment
     * @return {string} - HTML markup for the component
   */

  return (
    <button className='archive_button btn btn-danger' onClick={e => deleteComment(e)} type='submit'><FontAwesome name='trash' className='' /></button>
  );
};

/**
 * @method module:Reactator.ReduxContainerBuilderMapStateToProps
 *
 * @param {function} mapStateToProps - function mapping redux state to props
 *
 * @return {ReduxContainerBuilder} this builder
*/
const mapStateToProps = state => ({
  comment: state.comment,
});

export default connect(
  mapStateToProps,
  {
    deleteArticleComment, deleteCommentReplies,
  },
)(DeleteButton);
