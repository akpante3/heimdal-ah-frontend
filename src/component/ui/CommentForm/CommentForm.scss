@import '../../../styles/variables.scss';

.form__comment {
  font-family:$font_family_reem-kufi;
  border-top: 1px solid $heimdal-comment-border;
  
  .comment__head__a {
    font-size: 25px;
    margin-top: 60px;
    padding-top: 70px;
  }
  .comment__head__b {
    font-size: 20px;
  }
  .comment-form {
    border: 1px solid $heimdal-comment-border;
    border-radius: 0!important; 
    margin-right: 70px!important;
    width: 100%;
  }
  
  .flex-container__a {
    display: flex;
    justify-content: flex-end;
    margin-top: 10px!important;
  }
  .flex-container__b {
    display: flex;
    justify-content: space-between;
  }
  
  .comment_check {
    margin-right: 30px!important;
    margin-top: 8px;

  }

  .private__comment {
    margin-right: 50px;
    margin-top: 10px;
    font-size: 15px;
    color: $heimdal-private-comment-color;
    font-family: $font-family__cardo;
  }
}

