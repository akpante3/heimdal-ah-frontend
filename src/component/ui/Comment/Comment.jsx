/* eslint-disable react/require-default-props */
/* eslint-disable react/no-unused-prop-types */
/* eslint-disable react/forbid-prop-types */
/* eslint-disable import/no-named-as-default */
import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import CommentCard from '../CommentCard/CommentCard';
import './Comment.scss';
import CommentForm from '../CommentForm/CommentForm';
import { getArticleComment } from '../../../actions/CommentActions/CommentActions';


/**
  * renderComponent
  * @method Class Class based Component
  * @summary React component for rendering the card components
  * @param {object}  comment contains the comments of an article
  * @param {Object} props - React PropTypes
  * @property {String|Object} className - String className compatible object for styling
  * @return {Node} React node containing comment card view
  */
export class Comment extends Component {
  componentDidMount() {
    this.props.getArticleComment(1);
  }

  /**
     * Renders the component.
     *
     * @memberof app.components.Comment
     * @return {string} - HTML markup for the component
   */
  render() {
    return (
      <Fragment>
        <div className='comment-header'>
        Comment
        </div>
        <CommentCard loading={this.props.loading} />
        <div className='comment__form'>
          <CommentForm />
        </div>
      </Fragment>
    );
  }
}

Comment.propTypes = {
  getArticleComment: PropTypes.func.isRequired,
  comment: PropTypes.object.isRequired,
  loading: PropTypes.bool.isRequired,
  status: PropTypes.bool,
};

/**
 * @method module:Reactator.ReduxContainerBuilderMapStateToProps
 *
 * @param {function} mapStateToProps - function mapping redux state to props
 *
 * @return {ReduxContainerBuilder} this builder
*/
const mapStateToProps = state => ({
  comment: state.comment,
  loading: state.comment.loading,
  status: state.state,
});

export default connect(mapStateToProps, { getArticleComment })(Comment);
