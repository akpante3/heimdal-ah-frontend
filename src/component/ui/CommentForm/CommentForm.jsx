/* eslint-disable react/no-unused-prop-types */
/* eslint-disable react/forbid-prop-types */
import React, { Component } from 'react';
import { FormGroup, FormControl } from 'react-bootstrap';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Button from '../Buttons/Button';
import { postArticleComment } from '../../../actions/CommentActions/CommentActions';
import Alert from '../Alert/Alert';
import './CommentForm.scss';

/**
  * renderComponent
  * @method Class Class based Component
  * @summary React component for rendering the card components
  * @param {object}  comment contains the comments of an article
  * @param {Object} props - React PropTypes
  * @property {String|Object} className - String className compatible object for styling
  * @return {Node} React node containing comment form view
  */
export class CommentForm extends Component {
  state = {
    formIsValid: false,
    commentError: '',
    content: '',
    checkbox: false,
  }

   onChange = (e) => {
     this.setState({
       [e.target.name]: e.target.value,
       commentError: '',
     });
   }

   onBlur = (e) => {
     if (e.target.value === '') {
       this.setState({
         commentError: 'Content cannot be Empty',
       });
     } else {
       this.setState({
         formIsValid: true,
         commentError: '',
       });
     }
   }

   onCheck = () => {
     const { checkbox } = this.state;
     this.setState({
       checkbox: !checkbox,
       commentError: '',
     });
   }

   submitForm = (e) => {
     e.preventDefault();
     const { user } = this.props.auth;
     const { formIsValid, content, checkbox } = this.state;
     if (content.length >= 1500) {
       this.setState({
         formIsValid: false,
         commentError: "can't have more than 1500 characters",
       });
     } else if (formIsValid) {
       const data = {
         content,
         isPrivate: checkbox,
       };

       this.props.postArticleComment(this.props.articleId, data, user.image);
       this.setState({
         content: '',
       });
     } else {
       this.setState({
         commentError: 'Content cannot be Empty',
       });
     }
   }

   /**
     * Renders the component.
     *
     * @memberof app.components.Comment
     * @return {string} - HTML markup for the component
   */
   render() {
     const { checkbox, commentError } = this.state;
     return (
       <div className='container form__comment'>
         <div className='comment__head'>
           <p className='comment__head__a'>ADD COMMENT</p>
           <p className='comment__head__b'>Comment</p>
         </div>
         <div>
           {' '}
           {commentError !== '' && <Alert type='warning' message={this.state.commentError} title='Invalid Credenntials' /> }
           <div />
         </div>
         <form className='' onSubmit={this.submitForm}>
           <FormGroup controlId="formControlsTextarea">
             <FormControl
               onBlur={this.onBlur}
               onChange={this.onChange}
               componentClass="textarea"
               value={this.state.content}
               name='content'
               className='comment-form'
               rows="7"
               cols="70"
               placeholder="textarea"
             />
           </FormGroup>
           <div className='flex-container__a'>
             <div className="form-check flex-container__b">
               <div className='private__comment'>Make it Private</div>
               <div className='comment_check'>
                 <input type="checkbox" name='checkbox' onChange={e => this.onCheck(e)} className="form-check" value={!checkbox} />
               </div>
             </div>
             <Button type="comment" className='comment-button' label="Add Comment" />
           </div>
         </form>
       </div>
     );
   }
}

CommentForm.propTypes = {
  postArticleComment: PropTypes.func.isRequired,
  comment: PropTypes.object.isRequired,
  auth: PropTypes.object.isRequired,
  error: PropTypes.object.isRequired,
};

/**
 * @method module:Reactator.ReduxContainerBuilderMapStateToProps
 *
 * @param {function} mapStateToProps - function mapping redux state to props
 *
 * @return {ReduxContainerBuilder} this builder
*/
const mapStateToProps = state => ({
  comment: state.comment,
  auth: state.auth,
  error: state.errors,
});

export default connect(mapStateToProps, { postArticleComment })(CommentForm);
