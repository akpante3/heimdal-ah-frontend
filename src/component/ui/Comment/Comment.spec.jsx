/* eslint-disable no-undef */
/* eslint-disable react/react-in-jsx-scope */
import { shallow } from 'enzyme';
import { Comment } from './Comment';
import { CommentCard } from '../CommentCard/CommentCard';
import { CommentForm } from '../CommentForm/CommentForm';

describe('rendering', () => {
  let wrapper;
  let props;
  beforeEach(() => {
    props = {
      getArticleComment: jest.fn(),
      comment: '',
    };
  });

  it('invokes `componentDidMount` when mounted', () => {
    const spy = jest.spyOn(Comment.prototype, 'componentDidMount');
    wrapper = shallow(<Comment comment='' getArticleComment={() => {}} />);
    wrapper.instance().componentDidMount();
    expect(spy).toHaveBeenCalled();
  });

  it('renders One Comment Card components', () => {
    wrapper = shallow(<Comment {...props} />);
    expect(wrapper.find(CommentCard)).toEqual({});
  });

  it('renders One Comment Form components', () => {
    wrapper = shallow(<Comment {...props} />);
    expect(wrapper.find(CommentForm)).toEqual({});
  });

  it('should render the component', () => {
    wrapper = shallow(<Comment {...props} />);
    expect(wrapper.length).toBe(1);
  });

  it('should render the component', () => {
    wrapper = shallow(<Comment {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
