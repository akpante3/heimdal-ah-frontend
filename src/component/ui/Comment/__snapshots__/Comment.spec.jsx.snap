// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should render the component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Comment
    comment=""
    getArticleComment={
      [MockFunction] {
        "calls": Array [
          Array [
            1,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <div
          className="comment-header"
        >
          Comment
        </div>,
        <Connect(CommentCard) />,
        <div
          className="comment__form"
        >
          <Connect(CommentForm) />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Comment",
          "className": "comment-header",
        },
        "ref": null,
        "rendered": "Comment",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "loading": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Connect(CommentForm) />,
          "className": "comment__form",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <div
            className="comment-header"
          >
            Comment
          </div>,
          <Connect(CommentCard) />,
          <div
            className="comment__form"
          >
            <Connect(CommentForm) />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Comment",
            "className": "comment-header",
          },
          "ref": null,
          "rendered": "Comment",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "loading": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Connect(CommentForm) />,
            "className": "comment__form",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
